File: app/src/main/java/com/deepak/umber/data/model/Categories.kt
package com.deepak.umber.data.model

/**
 * The fixed label set for the classifier.
 *
 * [ALL] order IS the model's label encoding — index N in this list is output neuron N. Changing the
 * order or removing an entry silently corrupts every stored weight matrix, so any edit here MUST
 * bump [LABELS_VERSION], which forces the model to be discarded and retrained from confirmed
 * history on next launch.
 *
 * Appending to the end is also a label-set change (the output layer widens), so bump for that too.
 */
object Categories {

    const val LABELS_VERSION = 1

    const val FOOD = "Food & Dining"
    const val GROCERIES = "Groceries"
    const val TRANSPORT = "Transport"
    const val SHOPPING = "Shopping"
    const val BILLS = "Bills & Utilities"
    const val ENTERTAINMENT = "Entertainment"
    const val HEALTH = "Health"
    const val EDUCATION = "Education"
    const val RENT = "Rent & Housing"
    const val INVESTMENTS = "Investments"
    const val TRANSFERS = "Transfers"
    const val CASH = "Cash"
    const val TRAVEL = "Travel"
    const val INCOME = "Income"
    const val OTHER = "Other"

    val ALL: List<String> = listOf(
        FOOD,
        GROCERIES,
        TRANSPORT,
        SHOPPING,
        BILLS,
        ENTERTAINMENT,
        HEALTH,
        EDUCATION,
        RENT,
        INVESTMENTS,
        TRANSFERS,
        CASH,
        TRAVEL,
        INCOME,
        OTHER,
    )

    val COUNT: Int = ALL.size

    fun indexOf(category: String): Int = ALL.indexOf(category)

    fun labelAt(index: Int): String = ALL.getOrElse(index) { OTHER }

    fun isValid(category: String): Boolean = ALL.contains(category)
}

File: server/app/categories.py
"""Canonical category taxonomy — mirrors Categories.kt in the Android app exactly.

Order matters on the Android side (it's the classifier's label encoding); the server doesn't
train a model, but keeps the same order and exact strings so a category round-trips identically
in both directions.
"""

FOOD = "Food & Dining"
GROCERIES = "Groceries"
TRANSPORT = "Transport"
SHOPPING = "Shopping"
BILLS = "Bills & Utilities"
ENTERTAINMENT = "Entertainment"
HEALTH = "Health"
EDUCATION = "Education"
RENT = "Rent & Housing"
INVESTMENTS = "Investments"
TRANSFERS = "Transfers"
CASH = "Cash"
TRAVEL = "Travel"
INCOME = "Income"
OTHER = "Other"

ALL = [
    FOOD,
    GROCERIES,
    TRANSPORT,
    SHOPPING,
    BILLS,
    ENTERTAINMENT,
    HEALTH,
    EDUCATION,
    RENT,
    INVESTMENTS,
    TRANSFERS,
    CASH,
    TRAVEL,
    INCOME,
    OTHER,
]

ALL_SET = frozenset(ALL)


def is_valid(category: str) -> bool:
    return category in ALL_SET


DIRECTIONS = ("DEBIT", "CREDIT")

CHANNELS = (
    "UPI",
    "CARD",
    "ATM",
    "IMPS",
    "NEFT",
    "RTGS",
    "NETBANKING",
    "AUTOPAY",
    "WALLET",
    "UNKNOWN",
)

# Priority order, highest trust first. A user's own correction always wins; MEMORY sits at the
# same tier as USER because a merchant-memory match only exists because a user confirmed that
# merchant once. See docs/SYNC.md "Conflict rules".
CATEGORY_SOURCE_TIER = {
    "USER": 0,
    "MEMORY": 0,
    "DASHBOARD": 1,
    "REMOTE": 2,
    "MODEL": 3,
    "SEED": 3,
    "NONE": 3,
}

CATEGORY_SOURCES = tuple(CATEGORY_SOURCE_TIER.keys())


def beats(candidate_source: str, candidate_updated_at: int, existing_source: str, existing_updated_at: int) -> bool:
    """True if a row with candidate_source/updated_at should overwrite one with existing_source/updated_at."""
    candidate_tier = CATEGORY_SOURCE_TIER.get(candidate_source, 3)
    existing_tier = CATEGORY_SOURCE_TIER.get(existing_source, 3)
    if candidate_tier != existing_tier:
        return candidate_tier < existing_tier
    return candidate_updated_at >= existing_updated_at
